lib/devise-security/controllers/helpers.rb

```diff
@@ -40,71 +40,80 @@ def valid_security_question_answer?(resource, answer)
 
       # controller instance methods
 
-        private
-
-        # lookup if an password change needed
-        def handle_password_change
-          return if warden.nil?
-
-          if !devise_controller? && !ignore_password_expire? && !request.format.nil? && request.format.html?
-            Devise.mappings.keys.flatten.any? do |scope|
-              if signed_in?(scope) && warden.session(scope)['password_expired']
-                # re-check to avoid infinite loop if date changed after login attempt
-                if send(:"current_#{scope}").try(:need_change_password?)
-                  store_location_for(scope, request.original_fullpath) if request.get?
-                  redirect_for_password_change scope
-                  return
-                else
-                  warden.session(scope)[:password_expired] = false
-                end
+      private
+
+      # Called as a `before_action` on all actions on any controller that uses
+      # this helper. If the user's session is marked as having an expired
+      # password we double check in case it has been changed by another process,
+      # then redirect to the password change url.
+      #
+      # @note `Warden::Manager.after_authentication` is run AFTER this method
+      #
+      # @note Once the warden session has `'password_expired'` set to `false`,
+      #    it will **never** be checked again until the user re-logs in.
+      def handle_password_change
+        return if warden.nil?
+
+        if !devise_controller? &&
+           !ignore_password_expire? &&
+           !request.format.nil? &&
+           request.format.html?
+          Devise.mappings.keys.flatten.any? do |scope|
+            if signed_in?(scope) && warden.session(scope)['password_expired'] == true
+              if send(:"current_#{scope}").try(:need_change_password?)
+                store_location_for(scope, request.original_fullpath) if request.get?
+                redirect_for_password_change(scope)
+              else
+                warden.session(scope)['password_expired'] = false
               end
             end
           end
         end
+      end
 
-        # lookup if extra (paranoid) code verification is needed
-        def handle_paranoid_verification
-          return if warden.nil?
+      # lookup if extra (paranoid) code verification is needed
+      def handle_paranoid_verification
+        return if warden.nil?
```

Comment:
Hi @olbrich.  I'm trying to understand what this PR does to fix the bug, but I'm finding it difficult to identify what functionally changed.  Very respectfully, could I suggest making the formatting changes in another PR, so the bugfix here is as clear as possible?  P.S. I really appreciate the work you're doing the fix the bug, as I'm affected by it - thank you!

Reply:
@dan-jensen Yeah, I'll clean this up a bit when I work on it again on Wednesday.
BTW, we have been struggling with coming up with scenarios to reproduce the original problem, so if you have more details that would be appreciated.

You are also welcome to join us for a LiveShare coding session, I'll add a link here on Wednesday (2020-05-20) night at about 8pm Eastern.